analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs

```diff
@@ -5,14 +5,24 @@
 public class EqualityOnFloatingPoint
 {
     bool HalfEqual(Half first, Half second)
-        => first == second;    // FN
+        => first == second;    // Noncompliant {{Do not check floating point equality with exact values, use a range instead.}}
+    //           ^^
 
     bool NFloatEqual(NFloat first, NFloat second)
-        => first == second;    // FN
+        => first == second;    // Noncompliant
 
     bool IsEpsilon<T>(T value) where T : IFloatingPointIeee754<T>
-        => value == T.Epsilon; // FN
+        => value == T.Epsilon; // Noncompliant
+
+    bool IsPi<T>(T value) where T : IFloatingPointIeee754<T>
+        => value <= T.Pi && ((value >= T.Pi)); // Noncompliant
+
+    bool IsNotE<T>(T value) where T : IFloatingPointIeee754<T>
```

Comment:
what about 

- [NaN](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.ifloatingpointieee754-1.nan?view=net-7.0#system-numerics-ifloatingpointieee754-1-nan)
- [NegativeInfinity](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.ifloatingpointieee754-1.negativeinfinity?view=net-7.0#system-numerics-ifloatingpointieee754-1-negativeinfinity)
- [PositiveInfinity](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.ifloatingpointieee754-1.positiveinfinity?view=net-7.0#system-numerics-ifloatingpointieee754-1-positiveinfinity)

should equality be treated the same as Noncompliant? I imagine not?

Reply:
From the docs:

> Use the [IsNaN](https://learn.microsoft.com/en-us/dotnet/api/system.double.isnan?view=net-7.0) method to determine whether a value is not a number. The [Equality](https://learn.microsoft.com/en-us/dotnet/api/system.double.op_equality?view=net-7.0) operator considers two [NaN](https://learn.microsoft.com/en-us/dotnet/api/system.double.nan?view=net-7.0) values to be unequal to one another.

[Source](https://learn.microsoft.com/en-us/dotnet/api/system.double.nan?view=net-7.0#remarks)

For the others:

> Use [IsPositiveInfinity](https://learn.microsoft.com/en-us/dotnet/api/system.double.ispositiveinfinity?view=net-7.0) to determine whether a value evaluates to positive infinity.

> Use [IsNegativeInfinity](https://learn.microsoft.com/en-us/dotnet/api/system.double.isnegativeinfinity?view=net-7.0) to determine whether a value evaluates to negative infinity.

There is nothing mentioned for NegativeZero.

The "Is..." methods are defined in [INumberBase](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.inumberbase-1.isnan?view=net-7.0#system-numerics-inumberbase-1-isnan(-0)) and IFloatingPointIeee754 derives from INumberBase.

The only exception seems to be NegativeZero which doesn't have an `IsNegativeZero` counterpart.

Reply:
Fair point. In this case, the rule message will be confusing
```
Do not check floating point equality with exact values, use a range instead.
```
because for these values there are clear framework provided methods for doing the checks:
- IsNaN
- IsPositiveInfinity
- IsNegativeInfinity

And the rule should rather say that in the message. I'd say that there should be a ticket for this improvement.

Reply:
#6521